src/Microsoft.AspNet.Security.OpenIDConnect/OpenidConnectAuthenticationHandler.cs

```diff
@@ -0,0 +1,621 @@
+                ClaimsPrincipal principal = validator.ValidateToken(openIdConnectMessage.IdToken, validationParameters, out validatedToken);
+
+                JwtSecurityToken jwt = validatedToken as JwtSecurityToken;
+                if (jwt == null)
+                {
+                    throw new InvalidOperationException("Validated Security Token must be a JwtSecurityToken was: " + validatedToken.GetType());
+                }
+
+                AuthenticationTicket ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationType);
+                string nonce = RetrieveNonce(openIdConnectMessage);
+                if (!string.IsNullOrWhiteSpace(openIdConnectMessage.SessionState))
+                {
+                    ticket.Properties.Dictionary[OpenIdConnectSessionProperties.SessionState] = openIdConnectMessage.SessionState;
+                }
+
+                if (!string.IsNullOrWhiteSpace(_configuration.CheckSessionIframe))
+                {
+                    ticket.Properties.Dictionary[OpenIdConnectSessionProperties.CheckSessionIFrame] = _configuration.CheckSessionIframe;
+                }
+
+                if (Options.UseTokenLifetime)
+                {
+                    // Override any session persistence to match the token lifetime.
+                    DateTime issued = validatedToken.ValidFrom;
+                    if (issued != DateTime.MinValue)
+                    {
+                        ticket.Properties.IssuedUtc = issued;
+                    }
+
+                    DateTime expires = validatedToken.ValidTo;
+                    if (expires != DateTime.MinValue)
+                    {
+                        ticket.Properties.ExpiresUtc = expires;
+                    }
+
+                    ticket.Properties.AllowRefresh = false;
+                }
+
+                var securityTokenValidatedNotification = new SecurityTokenValidatedNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions>(Context, Options)
+                {
+                    AuthenticationTicket = ticket,
+                    ProtocolMessage = openIdConnectMessage
+                };
+
+                await Options.Notifications.SecurityTokenValidated(securityTokenValidatedNotification);
+                if (securityTokenValidatedNotification.HandledResponse)
+                {
+                    return securityTokenValidatedNotification.AuthenticationTicket;
+                }
+
+                if (securityTokenValidatedNotification.Skipped)
+                {
+                    return null;
+                }
+
+                var protocolValidationContext = new OpenIdConnectProtocolValidationContext
+                {
+                    AuthorizationCode = openIdConnectMessage.Code,
+                    Nonce = nonce,
+                };
+
+                Options.ProtocolValidator.Validate(jwt, protocolValidationContext);
+                if (openIdConnectMessage.Code != null)
```

Comment:
Basic access token retrieval has been a major demand. Is this something you could consider?